// scss-bundle/dist/bundle-cli.js
#!/usr/bin/env node
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const chokidar = require("chokidar");
// tslint:disable-next-line:no-require-imports
const debounce = require("lodash.debounce");
const Contracts = require("./contracts");
const arguments_1 = require("./arguments");
const launcher_1 = require("./launcher");
function resolveVerbosity(verbosity) {
    // Convert given value to an appropriate Verbosity enum value.
    // 'as any as number' is used because TypeScript thinks
    //  that we cast string to number, even though we get a number there
    return Contracts.Verbosity[verbosity];
}
function argumentsToConfig(argumentValues) {
    return {
        Destination: argumentValues.dest,
        Entry: argumentValues.entry,
        DedupeGlobs: argumentValues.dedupe,
        Verbosity: resolveVerbosity(argumentValues.verbosity),
        IncludePaths: argumentValues.includePaths,
        IgnoredImports: argumentValues.ignoredImports,
        ProjectDirectory: path.resolve(process.cwd(), argumentValues.project)
    };
}
function main(argumentValues) {
    return __awaiter(this, void 0, void 0, function* () {
        const config = argumentsToConfig(argumentValues);
        const isWatching = argumentValues.watch != null;
        const bundler = new launcher_1.Launcher(config);
        if (argumentValues.verbosity !== Contracts.Verbosity.None && (argumentValues.entry == null || argumentValues.dest == null)) {
            console.error("[Error] 'entry' and 'dest' are required.");
            process.exit(1);
        }
        if (argumentValues.verbosity !== Contracts.Verbosity.None && isWatching && argumentValues.watch === "") {
            console.error("[Error] 'watch' must be defined.");
            process.exit(1);
        }
        if (isWatching) {
            const onFilesChange = debounce(() => __awaiter(this, void 0, void 0, function* () {
                if (config.Verbosity === Contracts.Verbosity.Verbose) {
                    console.info("[Watcher] File change detected.");
                }
                yield bundler.Bundle();
                if (config.Verbosity === Contracts.Verbosity.Verbose) {
                    console.info("[Watcher] Waiting for changes...");
                }
            }), 500);
            chokidar.watch(argumentValues.watch).on("change", onFilesChange);
        }
        yield bundler.Bundle();
        if (isWatching && config.Verbosity === Contracts.Verbosity.Verbose) {
            console.info("[Watcher] Waiting for changes...");
        }
    });
}
main(arguments_1.argv);
